Added command line options for temperature/precipitation gamma correction curve.

esampson opened the pull request:
Also minor code cleanup.

Modified scatter plot to show gamma correction curve.

ftomassetti commented on worldengine/cli/main.py:
```diff
@@ -313,6 +314,14 @@ def main():
                              "If not provided, the default values will be used. \n" +
                             "[default = .059/.222/.493/.764/.927/.986/.998]",
                             metavar="#/#/#/#/#/#/#")
+    g_generate.add_argument('-gv', '--gamma-value', dest='gv', type=float,
```
We really need good documentation for these values :)

tcld replied:
Yes. When I hear "gamma value" the first thing I think of is something related to the brightness of the image-output.
I would maybe like to see a link to some page that explains this in more detail.

esampson replied:
That's because this is the same formula.

Y = X^gamma where both X and Y are in the range of 0-1.

ftomassetti commented on worldengine/cli/main.py:
```diff
@@ -441,10 +450,10 @@ def main():
     if args.temps:
         temps = args.temps.split('/')
-        for x in range(0,6):
+        for x in range(0,len(temps)):
```
should we verify the input is correct? (correct number of items, values in correct range, decreasing order)

tcld replied:
I would suggest to add an assert to test the length of the array (otherwise a user might see crashes much later in the code and not understand them). Regarding the value-order etc. I would just leave that to the responsibility of the user. Should we start working on a global settings-file in the future, we could add a class to handle all kinds of stupid inputs.

(This also affects line 468.)

ftomassetti replied:
I think that also issues in the order or the value (range 0..1) could lead to crashes or at least unpredictable behavior later. The best person for confirming/denying this is probably @esampson 

I do not think we should use an assert but instead handle the errors printing a nice message and closing the application with an error code.

esampson replied:
We do test the length of the array. We don't test order because there is a very slight possibility someone will want to input a series where parameter X is less than parameter X-1 (this would result in a zone being skipped).

ftomassetti replied:
mmm I see. I would in that case issue a warning, so that users who invert the order of items by mistake could realize it. If then they would like to do so... free to do it

ftomassetti commented on worldengine/world.py:
```diff
@@ -34,6 +35,8 @@ def __init__(self, name, width, height, seed, num_plates, ocean_level,
         self.step = step
         self.temps = temps
         self.humids = humids
+        self.gamma_curve = gamma_curve
```
we should serialize these values. I opened an issue for that (#174) as we can do that also after having merged this patch

tcld commented on tests/serialization_test.py:
```diff
@@ -11,7 +11,7 @@ def setUp(self):
 
     def test_pickle_serialize_unserialize(self):
-        w = world_gen("Dummy", 32, 16, 1, [.874, .765, .594, .439, .366, .124], [.941, .778, .507, .236, 0.073, .014, .002], step=Step.get_by_name("full"))
+        w = world_gen("Dummy", 32, 16, 1, step=Step.get_by_name("full"))
```
I like this a lot better. :)

tcld commented on worldengine/cli/main.py:
```diff
@@ -313,6 +314,14 @@ def main():
                             "[default = .059/.222/.493/.764/.927/.986/.998]",
                             metavar="#/#/#/#/#/#/#")
+    g_generate.add_argument('-gv', '--gamma-value', dest='gv', type=float,
+                        help="N = Gamma value for temperature/precipitation " +
```
What is ```N = ``` good for?

tcld commented on worldengine/cli/main.py:
```diff
@@ -313,6 +314,14 @@ def main():
+    g_generate.add_argument('-gv', '--gamma-value', dest='gv', type=float,
+                        help="N = Gamma value for temperature/precipitation " +
+                             "gamma correction curve. [default = %(default)s]",
+                        metavar="N", default='1.25')
```
Do you actively use the ```metavar```-name somewher?
(Also in line 324.)

As far as I understand (https://docs.python.org/3/library/argparse.html#metavar) you shouldn't make use of that parameter and the (much more useful) ```dest```-name will be used in case it is needed.

esampson replied:
Metavar regulates how the option is displayed when you use -help.

tcld commented:
I noticed something:
In the current master-branch https://github.com/Mindwerks/worldengine/blob/master/tests/data/data_generator.py is broken due to missing parameters. You probably overlooked that file in the PR preceding this one. It would be nice if you could double-check that it works with this PR. (And it would be nice if this PR then could be merged as soon as possible. ;) )

esampson commented:
Added a routine to print warnings. Added warnings for unusual but legal assignments for humidity and temperature ranges. Added error catching for gamma value and gamma offset.

tcld commented:
Good job on that front. :) I hope that we can at some point tackle "global variables" in a way that shortens that code a bit (or rather puts it somewhere else^^).

(Also, please don't overlook my last comment.)

esampson commented:
Ok. Fixed data_generator.py by adding defaults to _plates_simulation. Also added a coverage test for the new warnings systems.

ftomassetti commented on worldengine/cli/main.py:
```diff
@@ -482,6 +500,26 @@ def main():
+    if temps != sorted(temps, reverse=True):
+        warnings.append("WARNING: Temperature array not in ascending order")
+    if numpy.amin(temps) < 0:
+        warnings.append("WARNING: Maximum value in temperature array greater than 1")
```
perhaps the messages are inverted

esampson replied:
No, the warnings are actually correct. We invert the values that are given because I think it will be easier for people to understand and I worry that if I go through all of the code to handle the new values directly I'll end up missing bits and it will end up not working correctly.

ftomassetti commented on worldengine/plates.py:
```diff
@@ -36,25 +36,30 @@ def generate_plates_simulation(seed, width, height, sea_level=0.65,
-def world_gen(name, width, height, seed, temps, humids, num_plates=10, ocean_level=1.0,
-              step=Step.full(), fade_borders=True, verbose=get_verbose()):
+def world_gen(name, width, height, seed, temps=[.874, .765, .594, .439, .366, .124],
+              humids=[.941, .778, .507, .236, 0.073, .014, .002], num_plates=10,
```
could we use a constant to avoid repeating the default values for humids and temps?

ftomassetti commented:
Thanks for adding the warnings!

ftomassetti commented:
No idea why AppVeyor failed. Restarting it.

ftomassetti commented:
Apparently the last build was killed because it was taking too long (and it was taking too long during the setup phase). I do not see a way to restart the build.

psi29a commented:
Bumped it for a rebuild. First log-in, then you can get the rebuild button. :)

ftomassetti commented:
All the tests passing. Merging. Thanks to everyone!
